File: ruby/Visitor/general_report.rb
require_relative 'customer'
require_relative 'visitor'
require_relative 'order'

class GeneralReport < Visitor
  attr_accessor :customers_no
  attr_accessor :orders_no

  def visit(customer)
    if customer.kind_of? Customer
      print "Cusomer #{customer.name}\n"
      @customers_no ||= 0
      @customers_no = @customers_no + 1
    else
      print "Order #{customer.name}\n"
      @orders_no ||= 0
      @orders_no += 1
    end
  end

  def display_results
    print "Number of customers: #{customers_no}\n"
    print "Number of orders: #{orders_no}\n"
  end
end

File: java/Visitor/src/GeneralReport.java

public class GeneralReport implements Visitor{

	private int customersNo;
	private int ordersNo;

	public void visit(Customer customer)
	{
		System.out.println("Customer "+customer.name);
		customersNo ++;
	}
	public void visit(Order order)
	{
		System.out.println("Order "+order.name);
		ordersNo++;
	}

	public void displayResults()
	{
		System.out.println("Number of customers:" + customersNo);
		System.out.println("Number of orders:   " + ordersNo);
	}
}